Add accessibility guidelines for contributors (#17338)
django/django: docs/internals/contributing/writing-code/accessibility.txt

=============
Accessibility
=============

> All aspects of Django must be compatible with our accessibility standards,

Reviewer: ```suggestion
All aspects of Django must be compatible with our accessibility standards
```
As per the other PR I mentioned this with ;)

> and with the specific support targets we aim for.

Reviewer 2: I'm not sure about the suggested use of "The Django project" here. Perhaps just saying "Django is committed …" would be better, but I'm not well-versed in the established practice for how we refer to ourselves.

.. _accessibility-standards:

Accessibility standards
=======================

> Django aims for conformance with `WCAG 2.2`_ at the AA level.

Reviewer 3: "aims" feels like a bit of a cop-out, I'd prefer stronger language here.

Author: I’ll try!

Reviewer 2: There's a tension that's hard to rectify: We definitely want to use strong language, but we also have to be honest about the fact that we're not there yet. In my suggestion on the intro above, I chose "We work … to …" as wording that indicates we are actively doing something, not just aiming/aspiring/seeking. Perhaps that sort of change would be good here, as well.

Author: I’ve tried to make this stronger, without making it sound like we’re there yet.

> Django aims for conformance with `WCAG 2.2`_ at the AA level.

Reviewer 2: We should spell out this acronym on first usage.

```suggestion
Django aims for conformance with the `Web Content Accessibility Guidelines (WCAG), version 2.2`_, at the AA level.
```

> WCAG is the most well established standard for web accessibility. AA-level

Reviewer: ```suggestion
WCAG is the most established standard for web accessibility. AA-level
```
"most established" is sufficient 👍

Author: Thanks! English isn’t my native language so very much appreciate that kind of feedback.

> guidelines are the most common legal compliance target around the world.

Reviewer: Suggestion: Consider using "worldwide" which is more succinct & common when describing things 'around the world' 😊
```suggestion
guidelines are the most common legal compliance target worldwide.
```

> Wherever applicable, we also attempt to follow other best practices, such as:

Reviewer 2: Similar to Tom's comment on "Where possible", the "Wherever applicable" here could benefit from some elaboration. Or maybe it should just be dropped:

```suggestion
We also attempt to follow other best practices, such as:
```

* WCAG 2.2 AAA-level guidelines.

> * The upcoming `WCAG 3.0`_ guidelines.

Author: Update – I’m going to remove this unless others disagree, for the purpose of making this content more approachable (no concerns about there being multiple versions of WCAG to be aware of). WCAG 3.0 feels like it’s at least 5 to 10 years away at a minimum, we can add it back once it’s closer to ready.

> * `ATAG 2.0`_ guidelines for authoring tools.

Reviewer 2: ```suggestion
* `Authoring Tool Accessibility Guidelines (ATAG) 2.0`_  – guidelines for software and services that are used to produce web content.
```

.. note::
    Do you want to learn more about accessibility without diving straight into
    standards? We recommend `The A11Y Project`_,
    a community-driven effort to make digital accessibility easier.

> .. _WCAG 2.2: https://www.w3.org/TR/WCAG22/
> .. _WCAG 3.0: https://www.w3.org/TR/wcag-3.0/
> .. _ATAG 2.0: https://www.w3.org/TR/ATAG20/

Reviewer 2: ```suggestion
.. _Web Content Accessibility Guidelines (WCAG), version 2.2: https://www.w3.org/TR/WCAG22/
.. _WCAG 3.0: https://www.w3.org/TR/wcag-3.0/
.. _Authoring Tool Accessibility Guidelines (ATAG) 2.0: https://www.w3.org/TR/ATAG20/
```

.. _The A11Y Project: https://www.a11yproject.com/

.. _accessibility-support-targets:

Accessibility support targets
=============================

Beyond standards, we want to make sure Django actually works for people using
a wide range of input devices and assistive technologies.
The best way to do this effectively is to take accessibility considerations
into account as part of designing features. If in doubt, consult

> with users who rely on assistive technologies or with accessibility experts.

Reviewer 3: > in the spirit of a "shift-left" approach.

What happened to not using technical jargon in documentation, Thibaud? 🤔

> with users who rely on assistive technologies or with accessibility experts.

Reviewer 3: But also, all good points.

> with users who rely on assistive technologies or with accessibility experts.

Author: 😂 you got me

> with users who rely on assistive technologies or with accessibility experts.

Reviewer 4: I agree with the paragraph but as I mentioned in a PR-level comment, I think it's scary. I believe it should have links to concrete resources to quantify how many are enough "accessibility considerations" and who/where to reach out for help.

> with users who rely on assistive technologies or with accessibility experts.

Author: Also noting [Andrew’s comment here](https://github.com/django/django/pull/17338#issuecomment-1843410088) as I believe this will be the right place to address this (TBC):

> As for the guidelines around strictness - I think it is especially important to say the implied part out loud (that progressive improvement in small parts is better than demanding perfection). We don't want to discourage people from contributing, we want to encourage them to instead make things a little bit better every time.

This was a reply to [my prev comment](https://github.com/django/django/pull/17338#issuecomment-1842250001)

.. _accessibility-testing-baseline:

Accessibility testing baseline
------------------------------

>  Always test user interface changes with:

Reviewer 2: Space-o:

```suggestion
Always test user interface changes with:
```

* Keyboard-only navigation

> * Automated checks in the `Accessibility Insights`_ browser extension,
>   or an equivalent tool with the `Axe`_ checker.

Reviewer 2: Maybe slightly clearer?

```suggestion
* The `Accessibility Insights`_ browser extension's automated checks feature,
  or an equivalent tool with the `Axe`_ checker.
```

>   or an equivalent tool with the `Axe`_ checker.

Author: @nessita the wording as-is says "Accessibility Insights or equivalent tool with the Axe checker", so I don’t understand how you equate this to forcing Chrome on contributors? There are a lot of tools available with this checker (a browser extension for Firefox, lots of command line tools, etc).

Do you think we should link to specific options here then? Perhaps link to a list of products / integrations of Axe?

>   or an equivalent tool with the `Axe`_ checker.

Reviewer 4: I guess my point is that, upon reading this sentence, I'm concerned that this approach might be intimidating for new contributors, as it places the burden of finding an "equivalent tool with the Axe checker" squarely on their shoulders.  Does this make sense?

>   or an equivalent tool with the `Axe`_ checker.

Reviewer 5: The new contributor experience is tough and I get your point here - but I think this is fine

My reasons are that given roughly 65% of browser users use a chrome based browser, having an expectation that someone will test against a chrome based browser when making UI updates is (to me) not unreasonable
Our selenium tests also only run against chrome in the CI and (as a result) have some failures on firefox, so the chrome preference is an existing part of the Django contributing process.

In terms of the new contributor experience, the main message is roughly "no one makes a contribution to Django on their own, if you need help - ask". But I think that is true of all contributions and the start of this section says `Reach out to the Accessibility team for support and to coordinate testing.`. So I think we have done a reasonable effort to say this.
As we request the accessibility team to review PRs (and given these new guidelines we should be able to feel we could do this if we wanted), "someone" should have done this check (ideally the contributor and the reviewer but at least a reviewer)

As a future TODO, we can add some of this in the PR template around the UI changes (we currently tell them to add screenshots - but we can say "and test for...")

> Where appropriate, also test with:

Reviewer 3: I think we need to define (or at least give examples) when it's appropriate.

Author: Great idea, will do.

Reviewer 4: Big +1, otherwise the "when" is subject to interpretation and subsequent potential frustration.

* Touch-only navigation
* 400% browser zoom

> * forced-colors mode DevTools emulation

Reviewer 2: Suggest we name-check the most common use case, and also link to instructions for how to emulate.

```suggestion
* Forced-colors mode (e.g., Windows Contrast Themes)
  * `Can be emulated with Chrome/Edge dev tools`_
  
.. _Can be emulated with Chrome/Edge dev tools: https://devtoolstips.org/tips/en/emulate-forced-colors/
```

> .. note::

Reviewer 6: Please try to avoid "bare" notes and use `admonition` instead. This guideline is from [Guidelines for reStructuredText files](https://docs.djangoproject.com/en/stable/internals/contributing/writing-documentation/#guidelines-for-restructuredtext-files).

Author: Could you say more @felixxm? I chose not to use `admonition` because to me that means the content is a "warning" of sorts. I also didn’t understand that specific guideline as from what I could see the two rendered the same as far as the published docs are concerned.

If this is about the lack of a heading (is that what "bare" means?), I chose not to because though it visually looks like a heading, it’s not properly marked up as such for screen reader users.

---

So if we want to revisit this I think I’d rather just have the content within the main body copy of the docs and avoid admonitions / notes altogether.

Reviewer 6: If it's a warning use `warning`. Notes and admonitions are like side notes, so you theoretically don't need to read them to understand the topic. Titled admonitions are better because you don't have to read the entire note to check what it is about.

Author: Should now be addressed.

    How much should you test? Devise the UI with accessibility in mind,
    and the testing will only be needed as a final check. For more complex

>     interfaces, confer with other contributors to decide on testing targets.

Reviewer 2: I agree that it's a worthwhile thing to include.

.. _assistive-technologies:

Assistive technologies
----------------------

Where more extensive testing is appropriate,

> here are free assistive technologies we recommend testing with:

Reviewer: ```suggestion
here are <determiner> free assistive technologies we recommend testing with:
```
We just need a "determiner" here like "some" or "a few"… 👍

* `NVDA`_ on Windows with Firefox ESR
* `VoiceOver`_ on macOS with Safari
* `Narrator`_ on Windows with Microsoft Edge
* Windows Magnifier and macOS Zoom
* Windows Speech Recognition and macOS Voice Control
* VoiceOver on iOS, or TalkBack on Android

> * `Contrast themes`_ on Windows

Author: This is heavily based on https://docs.wagtail.org/en/stable/contributing/developing.html#accessibility-targets / https://www.gov.uk/service-manual/technology/testing-with-assistive-technologies.

For the more demanding changes we would want to test with _all_ of those things, but is burdensome due to requiring 4 different operating systems. At least there’s something for everyone!

Worth mentioning two possible options I chose to exclude for now:

- [Orca](https://help.gnome.org/users/orca/stable/index.html.en) for Linux. There’s 0 reported usage on [screen reader user surveys](https://webaim.org/projects/screenreadersurvey9/) and 0.5% reported on https://accessibility.blog.gov.uk/2016/11/01/results-of-the-2016-gov-uk-assistive-technology-survey/. I’ve never used it myself nor know anyone who has used it. I think it would be a good one to include _if_ we have anyone contributing in Django who is willing to test with it.
- [ChromeVox](https://support.google.com/chromebook/answer/7031755?hl=en-GB) on ChromeOS. There is some reported usage in survey but as far as I know there is no other way to test it than having a ChromeBook (the browser extension for Chrome isn’t a viable option). Again a good one to include _if_ we have anyone contributing in Django who is willing to test with it.

Reviewer 7: > Worth mentioning two possible options I chose to exclude for now:
> * [Orca](https://help.gnome.org/users/orca/stable/index.html.en) for Linux. There’s 0 reported usage on [screen reader user surveys](https://webaim.org/projects/screenreadersurvey9/) and 0.5% reported on https://accessibility.blog.gov.uk/2016/11/01/results-of-the-2016-gov-uk-assistive-technology-survey/. I’ve never used it myself nor know anyone who has used it. I think it would be a good one to include _if_ we have anyone contributing in Django who is willing to test with it.

I believe it is important to also include Orca for Linux in the list of technologies for many reasons, including:
- in the WebAIM survey it is specified that Orca is present among the technologies but has been merged into "others" (see [WebAIM: Screen Reader User Survey #5 Results](https://webaim.org/projects/screenreadersurvey5/#primary) and [WebAIM: Screen Reader User Survey #6 Results](https://webaim.org/projects/screenreadersurvey6/#primary) _"While they were not listed as response options for simplicity's sake, the survey comments indicate that Orca, SuperNova, and Speakup (among others) were common "Other" screen readers."_)
- Linux is the most used operating system among Django developers with 41% (see [Django Developers Survey 2022 Results](https://lp.jetbrains.com/django-developer-survey-2022/#circular-chart-685)), which makes them a decidedly different user base than that of the aforementioned survey, where it is only 1.4% (see [WebAIM: Screen Reader User Survey #9 Results](https://webaim.org/projects/screenreadersurvey9/#os))
- it could be an excellent opportunity to enrich future surveys for Django, Python or even Wgatail by asking actual developers what assistive technologies they use
- I have only tried to use Orca, but if we don't find other users of it in the accessibility team I could help you with this activity

Author: Thanks @pauloxnet! Those sound like excellent reasons to me, I’ll add it unless others raise objections.

Where expert help is available, also consider testing with:

* `JAWS`_ on Windows

> * `Dragon`_ on Windows

Author: These feel essential to me because they are very popular, but I chose to keep them separate because:

- Their licenses cost a lot of money
- They’re Windows-only

.. _Accessibility Insights: https://accessibilityinsights.io/
.. _Axe: https://github.com/dequelabs/axe-core
.. _NVDA: https://www.nvaccess.org/
.. _Narrator: https://support.microsoft.com/en-us/windows/complete-guide-to-narrator-e4397a0d-ef4f-b386-d8ae-c172f109bdb1
.. _VoiceOver: https://support.apple.com/en-gb/guide/voiceover-guide/welcome/web
.. _TalBack: https://support.google.com/accessibility/android/answer/6283677
.. _Contrast themes: https://support.microsoft.com/en-us/windows/change-color-contrast-in-windows-fedc744c-90ac-69df-aed5-c8a90125e696
.. _JAWS: https://www.freedomscientific.com/products/software/jaws/
.. _Dragon: https://www.nuance.com/en-gb/dragon.html

.. _accessibility-known-issues-how-to-help:

Known issues and how to help
============================

> Parts of Django do not meet our accessibility targets at the moment.

Reviewer: Consider prefixing this sentence with "Some" or "There are some", eg:
```suggestion
There are some parts of Django that do not meet our accessibility targets at the moment.
```

We actively work on fixing issues both as part of ongoing maintenance,

> and bigger overhauls. To learn about known issues, and get involved, see:

Author: I thought it was important to acknowledge that Django is quite far from meeting all the goals stated above (it’ll be plenty obvious to people going through the Django admin with our stated standards and testing targets in particular) – but that’s something we’re working on (and _you_ can get involved 😉 ).

Reviewer 8: Maybe a link to a board of open issues would be cool, such that possible maintaners can _jump in_

Author: There are links to open issues in items 2 and 3 in the list below, is that different from what you’re expecting?

> * ``#accessibility`` on the `Django Discord server`_.

Reviewer 3: Nice.

=============
Accessibility
=============

The Django project is committed to being accessible to everyone
in all possible ways.
We work to conform to established accessibility standards,
and also to meet our own targets for supporting specific input devices

> and assistive technologies.

Author: I’ll rewrite this whole document to wrap exactly at 79 columns for every line. For the record, I find this really problematic for readability of the documentation, so in the future I’ll raise an issue to change this rule (probably after #17340).

.. _accessibility-standards:

Accessibility standards
=======================

We work to conform with the `Web Content Accessibility Guidelines`_ (WCAG),
version 2.2, at the AA level.
WCAG is the most established standard for web accessibility.
AA-level guidelines are the most common legal compliance target worldwide.

We also attempt to follow other best practices, such as:

* WCAG 2.2 AAA-level guidelines.
* The upcoming `WCAG 3.0`_ guidelines.
* `Authoring Tool Accessibility Guidelines`_ (ATAG) 2.0 – guidelines
  for software and services that are used to produce web content.

To learn more about accessibility without diving straight into standards,

> we recommend `The A11Y Project`_, a community-driven effort

Reviewer 8: An acronym definition of A11Y would be good here imo

Author: Hmmm interesting. Since this is the name of the website as they write it, I was hesitant to spell it out.

to make digital accessibility easier.

.. _Web Content Accessibility Guidelines: https://www.w3.org/TR/WCAG22/
.. _WCAG 3.0: https://www.w3.org/TR/wcag-3.0/
.. _Authoring Tool Accessibility Guidelines: https://www.w3.org/TR/ATAG20/
.. _The A11Y Project: https://www.a11yproject.com/

.. _accessibility-support-targets:

> Accessibility support targets

Reviewer 4: ```suggestion
Support scope
```

Author: Could you explain your suggestion? I have spent a lot of time on wording choices here, and based this on other existing guidelines that use this wording.

=============================

Beyond standards, we want to make sure Django actually works for people using
a wide range of input devices and assistive technologies. This has benefits
to make sure Django works better `for everyone`_.
The best way to do this effectively is to take accessibility considerations
into account as part of designing features. If in doubt, consult
with users who rely on assistive technologies or with accessibility experts.

.. _for everyone: https://en.wikipedia.org/wiki/Curb_cut_effect

.. _accessibility-testing-baseline:

> Accessibility testing baseline

Reviewer 4: ```suggestion
Testing baseline
```

Reviewer 4: Even if someone lands in a heading, the topic of the document is "Accessibility". Having this prefixed on every section title feels redundant and I think it's fair to expect people to be able to scroll up and down for context if they need it.

------------------------------

Design the UI with accessibility in mind,
and the testing will only be needed as a final check. For more complex
interfaces, confer with other contributors to decide on testing targets.

> Reach out to the `Accessibility team`_ for support and to coordinate testing

Reviewer 4: Remember the wrapping at 79 cols :-)

Always test user interface changes with:

> * Keyboard-only navigation

Reviewer 9: Is there some resource that we could link to that gives a checklist for what to cover here? 

I can tab about, etc., but likely there’s lots that I wouldn’t think or know consider.

Author: Good point, yes definitely! The TL;DR; is you should be able to do everything with just tab, arrow keys, space and enter. There are some more nuances which a resource will be helpful for.

* The `Accessibility Insights`_ browser extension's automated checks feature,
  or an equivalent tool with the `Axe`_ checker.

Where appropriate, also test with:

* Touch-only navigation
* 400% browser zoom

> * Forced-colors mode (for example Windows Contrast Themes)

Reviewer 4: I would love to see each one of these be a link to a How-To provided by us.

Reviewer 4: Just to clarify my point here, I think we need to describe somewhere how to achieve forced-colors mode (I'm honestly asking out of ignorance). Is it a browser setting? Is it browser specific? Is it an operating system setting?

.. _assistive-technologies:

Assistive technologies
----------------------

> Where more extensive testing is appropriate,

Reviewer 4: Going back to @knyghty's point, we should make explicit where is appropriate.

here are popular free assistive technologies we recommend testing with:

> * `NVDA`_ on Windows with Firefox ESR

Reviewer 4: Does this mean that these tools are *not* available for non-ESR Firefox?

Author: No, it just means we recommend testing with Firefox ESR.

* `VoiceOver`_ on macOS with Safari
* `Orca`_ on Linux with Firefox ESR
* `Narrator`_ on Windows with Microsoft Edge
* Windows Magnifier and macOS Zoom
* Windows Speech Recognition and macOS Voice Control
* VoiceOver on iOS, or TalkBack on Android
* `Contrast themes`_ on Windows

> Where expert help is available, also consider testing with:

Reviewer 4: I'm not sure what this means :thinking:

Author: I’ll rephrase. This means "If you have the expertise or you can work with someone who does, also test with:".

make sure Django works better `for everyone`_. The best way to do this
effectively is to take accessibility considerations into account as part of
designing features. If in doubt, consult with users who rely on assistive
technologies or with accessibility experts.

.. _for everyone: https://en.wikipedia.org/wiki/Curb_cut_effect

.. _accessibility-testing-baseline:

Accessibility testing baseline
------------------------------

Design the UI with accessibility in mind, and the testing will only be needed
as a final check. For more complex interfaces, confer with other contributors
to decide on testing targets. Reach out to the `Accessibility team`_ for
support and to coordinate testing.

Always test user interface changes with:

* Keyboard-only navigation
* The `Accessibility Insights`_ browser extension's automated checks feature,
  or an equivalent tool with the `Axe`_ checker.

Where the UI changes could affect those modalities, also test with:

* Touch-only navigation
* 400% browser zoom
* Forced-colors mode (for example Windows Contrast Themes)

.. _assistive-technologies:

Assistive technologies
----------------------

> Where the UI changes could affect assistive technologies, here are popular
> free options we recommend testing with:
> 
> * `NVDA`_ on Windows with Firefox ESR
> * `VoiceOver`_ on macOS with Safari
> * `Orca`_ on Linux with Firefox ESR
> * `Narrator`_ on Windows with Microsoft Edge
> * Windows Magnifier and macOS Zoom
> * Windows Speech Recognition and macOS Voice Control
> * VoiceOver on iOS, or TalkBack on Android
> * `Contrast themes`_ on Windows

Reviewer 5: ```suggestion
Where the UI changes could affect assistive technologies, here are popular free
options we recommend testing with.

For Windows:

* `NVDA`_ - recommended with Firefox ESR
* `Narrator`_ - recommended with Microsoft Edge
* Windows Magnifier
* Windows Speech Recognition
* `Contrast themes`_

For macOS:

* `VoiceOver`_ - recommended with Safari
* macOS Zoom
* macOS Voice Control

For Linux:

* `Orca`_ - recommended with Firefox ESR

For mobile or tablet:

* VoiceOver on iOS, or TalkBack on Android
```
Suggestion: the reader will probably only have access to one OS and so I would group this and I tried to incorporate [this comment](https://github.com/django/django/pull/17338/files#r1487963102)
I am not sure on the "mobile or tablet" group name
